rule.py

```diff
@@ -1,73 +1,154 @@
+            return True 
+        else : return False 
     
-    def elephant(self, old_pos, new_pos):
-        pass 
+    def is_in_lower_half(self, new_pos):
+        if ( pos.x >= self._DOMAIN_X_LOWER_HALF[0] and pos.x <= self._DOMAIN_X_LOWER_HALF[1]) and ( pos.y >= self._DOMAIN_Y_LOWER_HALF[0] and pos.y <= self._DOMAIN_Y_LOWER_HALF[1]):
+            return True 
+        else : return False 
 
-    def advisor(self, old_pos, new_pos):
-        pass 
+    def is_in_northern_palace(self, new_pos):
+        if ( pos.x >= self._DOMAIN_Y_NORTHERN_PALACE[0] and pos.x <= self._DOMAIN_X_NORTHERN_PALACE[1]) and ( pos.y >= self._DOMAIN_Y_NORTHERN_PALACE[0] and pos.y <= self._DOMAIN_Y_NORTHERN_PALACE[1]):
+            return True 
+        else : return False 
 
-    def general(self, old_pos, new_pos):
-        pass 
+    def is_in_southern_palace(self, new_pos):
+        if ( pos.x >= self._DOMAIN_X_SOUTHERN_PALACE[0] and pos.x <= self._DOMAIN_X_SOUTHERN_PALACE[1]) and ( pos.y >= self._DOMAIN_Y_SOUTHERN_PALACE[0] and pos.y <= self._DOMAIN_Y_SOUTHERN_PALACE[1]):
+            return True 
+        else : return False 
+
+    def you_shall_not_pass(self, old_pos, new_pos):
+        imperial_guard = ['elephant', 'advisor', 'general']
+        the_one = self.get_piece(old_pos).type 
+        if the_one in imperial_guard: 
+            the_origin_coords = the_one.history[-1]            
+            if the_one == imperial_guard[0]: 
+                if self.is_in_lower_half(the_origin_coords): 
+                    return self.is_in_lower_half(new_pos)
+                elif self.is_in_upper_half(the_origin_coords): 
+                    return self.is_in_upper_half(new_pos)
+                else: 
+                    raise ValueError('The Elephant is travelling oversea!!! MovRuleSpecial_1: Chesspiece is not detected in board')
+            elif the_one == imperial_guard[1]: 
+                if self.is_in_northern_palace(the_origin_coords):
+                    return self.is_in_northern_palace(new_pos)
+                elif self.is_in_southern_palace(the_origin_coords):
+                    return self.is_in_southern_palace(new_pos)
+                else: 
+                    raise ValueError('Advisor not detected in the palace!!! MovRuleSpecial_2: Chesspiece is not detected in both palaces')
+            elif the_one == imperial_guard[2]: 
+                if self.is_in_southern_palace(the_origin_coords):
+                    return self.is_in_southern_palace(new_pos)
+                elif self.is_in_northern_palace(the_origin_coords): 
+                    return self.is_in_northern_palace(new_pos)
+                else: 
+                    raise ValueError('This is bad, the General is missing in the palace!!! MovRuleSpecial_3: Chesspiece is not detected in both palaces')
+            else: 
+                raise ValueError('What is the world is this thing? Identify thyself!! MovRuleSpecial_4: Chesspiece does not belong to the special group')
+
```

Comment:
Hmm, I wonder if you should check all the not passing rule in one function like this or not. In my opinion, it would be better to check the piece not passing rule separately in the function that the piece belongs to in the moving rule class.

Reply:
This one just essentially return true if the move is acceptable or false if not and error if none of the above. It did not touch anything else. the moving rule class is the class in the rule.py or a whole new class somewhere else?

Reply:
Ah, no. There is nothing wrong with this. I just give my opinion on this.